core/src/main/java/io/grpc/internal/ManagedChannelImpl2.java

```diff
@@ -1332,9 +1333,13 @@ public void run() {
           ConfigOrError configOrError = resolutionResult.getServiceConfig();
           ServiceConfigHolder validServiceConfig = null;
           Status serviceConfigError = null;
+          Map<String, ?> rawServiceConfig =
+              resolutionResult.getAttributes().get(GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG);
+          if (configOrError == null && rawServiceConfig != null) {
+            // use service config json in deprecated attributes
+            configOrError = serviceConfigParser.parseServiceConfig(rawServiceConfig);
+          }
           if (configOrError != null) {
```

Comment:
I think the original code is correct.

For a LoadBalancer that's still on the old service config path, it won't override `LoadBalancerProvider.parseLoadBalancingPolicyConfig`, thus `configOrError` would be `LoadBalancerProvider.UNKNOWN_CONFIG`, which is non-null, thus `NAME_RESOLVER_SERVICE_CONFIG` will get populated. We shouldn't call `parseServiceConfig` again.

Reply:
Some 1st/3rd party name resolver might not use the `serviceConfig` field which can be ignored in current implementation even if the deprecated `Attribute` is set. 
`LoadBalancerProvider.UNKNOWN_CONFIG` is still a valid config from the name resolver point of view, thus those aren't consumed.

Reply:
Now I see this is for "old" `NameResolver`s that don't call `parseServiceConfig()` at all but emit the raw config. Granted the raw config attribute is defined in the internal `GrpcAttributes` thus no third-party NameResolvers are expected to do that. It's fine to call parser here if that makes the migration for first-party NameResolver easier. It's going to be removed along with `GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG` anyway so not a big deal.

Reply:
it turns out that internal migration was easy. Thus, as pointed out the attributes shouldn't be used out side of gRPC. removing the service config parsing in `ManagedChannelImpl2`.